### PR TITLE
PLT-6018: Add APIv4 Endpoint for fetching deleted Channels

### PR DESCRIPTION
**Pull request by senk**
#### Summary
Adds APIv4 Endpoint to get deleted Channels
#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-6018
#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs) https://github.com/mattermost/mattermost-api-reference/pull/190
- [x] All new/modified APIs include changes to the drivers

### CODE REVIEW

**Comment by jwilander** [comment]
Looks good overall @senk, just a couple comments:

1. Can we add this endpoint API v4 instead of v3? We're trying to refrain from adding new endpoints to v3 and to use v4 when possible. When doing this you'll need to update the handler to use the v4 patterns
2. Instead of using `deleted`, can we use `archived`? We think of soft deletes as archiving and it'd be nice if the internal code and endpoint matched how we're going to talk about the feature

**Comment by senk** [comment]
1. Will be implemented as part of https://mattermost.atlassian.net/browse/PLT-6018
2. There is e.g a method ["GetDeletedByName"](https://github.com/mattermost/platform/blob/master/store/sql_channel_store.go#L790) so maybe we stick to _deleted_ for consistency reasons? the method in API v3 to _archive_ a channel is also _delete_..

**Comment by jwilander** [comment]
1. We should just add this to v4, not to v3, regardless of what that ticket says. There's no reason to add it to v3 and v4 both since we'll be making the full switch over to only use v4 in the next month or two. @grundleborg can chime in if he feels strongly about it, but less code to maintain is better
2. That's fine, if we're already using `deleted` internally we can continue with that

**Comment by grundleborg** [comment]
I'm going to agree with @jwilander on this. Since we've started moving to api4 only for other things now, there's not much point adding this to apiv3. @senk would you mind changing the api parts of this to be api4 instead? You can keep targeting this PLT or switch it to PLT-6018 - I don't really mind which (and I'll close the other one).

**Comment by grundleborg** [comment]
Also, @jwilander APIv4 would be a good point to think about whether we want to rename these endpoints to `archive` instead of `delete` since we it isn't stable yet.

**Comment by jasonblais** [comment]
Let's also close [APIv3 endpoint for Unarchiving Channels](https://mattermost.atlassian.net/browse/PLT-6017) in favor of [the corresponding APIv4 endpoint](https://mattermost.atlassian.net/browse/PLT-6019)

**Comment by grundleborg** [comment]
@jasonblais done.

**Comment by jwilander** [comment]
I'm rater 0/5 on `archived` vs `delete`. I'm 4/5 that if we choose one we go change all the references to delete to archive. My only reservation of going to archive is that we archive channels, but do we archive posts, teams, etc.? I only really care about it being consistent

**Comment by senk** [comment]
When we change every reference to delete to archive we should for consistency also rename the database column from _deleted_at_ to _archived_at_ . I don't know the side-effects of that..
@grundleborg @jasonblais I will target future commits to PLT-6019

**Comment by grundleborg** [comment]
OK, let's stick with `delete` for the name for now then. It'll get messy if we start distinguishing between `archive` and `delete` for different data types, when at an architectural level, it is really all the same thing.

**Comment by jasonblais** [comment]
Hey @senk, how are you doing on this PR? Anything that's blocking work or that we can help with?

Big thanks for your interested implementing the unarchived channels feature.

[Archived channels is a popular feature proposal](https://mattermost.uservoice.com/forums/306457-general/suggestions/10960347-channel-archiving) and this let's us get one step closer to achieving it. :)

**Comment by jwilander** [comment]
+1 on @grundleborg's comment

**Comment by senk** [comment]
Added Paging /cc @jwilander @grundleborg

File: api4/channel.go [hunk 1]
```diff
@@ -385,6 +386,26 @@ func getPublicChannelsForTeam(c *Context, w http.ResponseWriter, r *http.Request
...
+	if c.Err != nil {
+		return
+	}
+
+	if !app.SessionHasPermissionToTeam(c.Session, c.Params.TeamId, model.PERMISSION_MANAGE_TEAM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_TEAM)
+		return
+	}
+
+	if channels, err := app.GetDeletedChannels(c.Params.TeamId, c.Params.Page, c.Params.PerPage); err != nil {
```
Comment (grundleborg) [inline api4/channel.go]: I think the second argument should be `c.Params.Page * c.Params.PerPage` in order to convert from page/per-page to offset/limit.

Comment (senk) [inline api4/channel.go]: Ah thanks! Fixed it :)

**Comment by grundleborg** [comment]
@senk you've got a GOFMT error, but other than that, the fix looks good.

**Comment by senk** [comment]
style error should be fixed, thanks

**Comment by jwilander** [comment]
@senk you might need to update this unit test:
```
--- FAIL: TestChannelStoreGetDeleted (0.01s)
	sql_channel_store_test.go:592: wrong list length
```

**Comment by senk** [comment]
updated the unit test @grundleborg @jwilander

**Comment by grundleborg** [comment]
@senk Thanks, that's looking great now. We're going to hold of merging this for a few days until we have branched for 3.9 development, but as soon as that's done, it's ready to merge.

**Pull request merged**